File: data-structures/pilha/pilha.js
/* Classe pilha é intuitiva; pode adicionar ao pico e apenas retirar do pico; exemplos de 
uso se relacionam com a possibilidade de retornar um evento de cada vez: Ctrl + Z; o histórico
do Google... 
Stack/Pilha*/ 


class Pilha {
    constructor() {
        this.itens = [];
    }

    empilhar(elemento) {
        this.itens.push(elemento);
    }

    desempilhar() {
        if (this.itens.length === 0) {
            return "A pilha está vazia!";
        }
        this.itens.pop();
    }

    topo() {
        if (this.itens.length === 0) {
            return "A pilha está vazia!";
        }

        return this.itens[this.itens.length -1];
    }

    tamanho() {
        return this.itens.length;
    }

    exibir() {
        console.log(this.itens.join(", "));
    }
}

let pilha = new Pilha();

pilha.empilhar("prato_1");
pilha.empilhar("prato_2");
pilha.empilhar("prato_3");
pilha.exibir();
pilha.desempilhar();
pilha.exibir();
let tamanho = pilha.tamanho();
console.log(tamanho);
let topo = pilha.topo();
console.log(topo);

File: data-structures/stack/pilha.cpp
/* FILO: first in last out.
Classe pilha é intuitiva; pode adicionar ao pico e apenas retirar do pico; exemplos de 
uso se relacionam com a possibilidade de retornar um evento de cada vez: Ctrl + Z; o histórico
do Google... 
Stack/Pilha*/ 

#include <iostream>
#include <stack>

int main() {
    std::stack<int> pilha;

    pilha.push(20);
    pilha.push(30);
    pilha.push(40);

    int topo = pilha.top();
    bool yes = pilha.empty();
    int tamanho = pilha.size();

    while (!pilha.empty()) {
        std::cout << pilha.top() << std::endl;
        pilha.pop();
    }
    std::cout << "Topo: " << topo << std::endl;
    
    if (yes) {
        std::cout << "Yes" << std::endl;

    } else {
        std::cout << "No" << std::endl;
    }

    std::cout << "Tamanho: " << tamanho << std::endl;
}